proparse/src/main/java/org/prorefactor/proparse/antlr4/Lexer.java

```diff
@@ -0,0 +1,1117 @@
+      appendToEOL();
+      // We have to do the define *before* getting next char.
+      macroDefine(PreprocessorParser.AMPGLOBALDEFINE);
+      getChar();
+      return makeToken(PreprocessorParser.AMPGLOBALDEFINE);
+    }
+    if ("&scoped-define".startsWith(macroType) && macroType.length() >= 4) {
+      appendToEOL();
+      // We have to do the define *before* getting next char.
+      macroDefine(PreprocessorParser.AMPSCOPEDDEFINE);
+      getChar();
+      return makeToken(PreprocessorParser.AMPSCOPEDDEFINE);
+    }
+
+    if ("&undefine".startsWith(macroType) && macroType.length() >= 5) {
+      // Append whitespace between &UNDEFINE and the target token
+      while (Character.isWhitespace(currChar)) {
+        append();
+        getChar();
+      }
+      // Append the target token
+      while ((!Character.isWhitespace(currChar)) && currInt != EOF_CHAR) {
+        append();
+        getChar();
+      }
+      // &UNDEFINE consumes up to *and including* the first whitespace
+      // after the token it undefines.
+      // At least that seems to be what Progress is doing.
+      if (currChar == '\r') {
+        append();
+        getChar();
+        if (currChar == '\n') {
+          append();
+          getChar();
+        }
+      } else if (currInt != EOF_CHAR) {
+        append();
+        getChar();
+      }
+      macroUndefine();
+      return makeToken(PreprocessorParser.AMPUNDEFINE);
+    }
+
+    if (macroType.equals("&analyze-suspend")) {
+      appendToEOL();
+      getChar();
+      return makeToken(PreprocessorParser.AMPANALYZESUSPEND);
+    }
+    if (macroType.equals("&analyze-resume")) {
+      appendToEOL();
+      getChar();
+      return makeToken(PreprocessorParser.AMPANALYZERESUME);
+    }
+    if (macroType.equals("&message")) {
+      appendToEOL();
+      getChar();
+      return makeToken(PreprocessorParser.AMPMESSAGE);
+    }
+
+    if (macroType.equals("&if")) {
+      return makeToken(PreprocessorParser.AMPIF);
+    }
+    if (macroType.equals("&then")) {
+      return makeToken(PreprocessorParser.AMPTHEN);
+    }
+    if (macroType.equals("&elseif")) {
+      return makeToken(PreprocessorParser.AMPELSEIF);
+    }
+    if (macroType.equals("&else")) {
+      return makeToken(PreprocessorParser.AMPELSE);
+    }
+    if (macroType.equals("&endif")) {
+      return makeToken(PreprocessorParser.AMPENDIF);
+    }
+
+    // If we got here, it wasn't a preprocessor directive,
+    // and the caller is responsible for building the token.
+    return null;
+
+  }
+
+  //////////////// End lexical productions, begin support functions
+
+  void append() {
+    currText.append((char) currInt);
+  }
+
+  void append(char c) {
+    currText.append(c);
+  }
+
+  void append(String theText) {
+    currText.append(theText);
+  }
+
+  void appendToEOL() throws IOException {
+    // As with the other "append" functions,
+    // the caller is responsible for calling getChar() after this.
+    for (;;) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce the total number of break and continue statements in this loop to use at most one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://sonar.riverside-software.fr/coding_rules#rule_key=squid%3AS135)